src/ledger/device.cpp

```diff
@@ -0,0 +1,1679 @@
+      this->hCard = 0;
+    }
+
+    return true;
+  }
+
+  unsigned int Device::exchange(unsigned int ok, unsigned int mask) {
+    LONG rv;
+    int sw;
+
+    logCMD();
+
+    this->length_recv = BUFFER_SEND_SIZE;
+    rv = SCardTransmit(this->hCard,
+    SCARD_PCI_T0, this->buffer_send, this->length_send,
+    NULL,         this->buffer_recv, &this->length_recv);
+    ASSERT_RV(rv);
+
+    logRESP();
+    sw = (this->buffer_recv[this->length_recv-2]<<8) | this->buffer_recv[this->length_recv-1];
+    ASSERT_SW(sw,ok,msk);
+    return sw;
+  }
+
+  void Device::reset_buffer() {
+    this->length_send = 0;
+    memset(this->buffer_send, 0, BUFFER_SEND_SIZE);
+    this->length_recv = 0;
+    memset(this->buffer_recv, 0, BUFFER_SEND_SIZE);
+  }
+
+  /* ======================================================================= */
+  /*                                   MISC                                  */
+  /* ======================================================================= */
+  bool Device::reset() {
+    int offset;
+
+    lock_device();
+    reset_buffer();
+
+    this->buffer_send[0] = 0x00;
+    this->buffer_send[1] = INS_RESET;
+    this->buffer_send[2] = 0x00;
+    this->buffer_send[3] = 0x00;
+    this->buffer_send[4] = 0x00;
+    offset = 5;
+    //options
+    this->buffer_send[offset] = 0x00;
+
+    this->buffer_send[4] = offset-5;
+    this->length_send = offset;
+    this->exchange();
+    unlock_device();
+    return true;
+  }
+
+  /* ======================================================================= */
+  /*                             WALLET & ADDRESS                            */
+  /* ======================================================================= */
+
+  /* Application API */
+   bool Device::get_public_address(cryptonote::account_public_address &pubkey){
+    int offset;
+
+    lock_device();
+    reset_buffer();
+
+    this->buffer_send[0] = 0x00;
+    this->buffer_send[1] = INS_GET_KEY;
+    this->buffer_send[2] = 0x01;
+    this->buffer_send[3] = 0x00;
+    this->buffer_send[4] = 0x00;
+    offset = 5;
+    //options
+    this->buffer_send[offset] = 0x00;
+    offset += 1;
+
+    this->buffer_send[4] = offset-5;
+    this->length_send = offset;
+    this->exchange();
+
+    memmove(pubkey.m_view_public_key.data, this->buffer_recv, 32);
+    memmove(pubkey.m_spend_public_key.data, this->buffer_recv+32, 32);
+    unlock_device();
+    return true;
+  }
+
+  bool  Device::get_secret_keys(crypto::secret_key &viewkey , crypto::secret_key &spendkey) {
+    #ifdef DEBUGLEDGER
+    int offset;
+
+    lock_device();
+    reset_buffer();
+
+    this->buffer_send[0] = 0x00;
+    this->buffer_send[1] = INS_GET_KEY;
+    this->buffer_send[2] = 0x02;
+    this->buffer_send[3] = 0x00;
+    this->buffer_send[4] = 0x00;
+    offset = 5;
+    //options
+    this->buffer_send[offset] = 0x00;
+    offset += 1;
+
+    this->buffer_send[4] = offset-5;
```

Comment:
This writes into offset 4 again, instead of 6, over-writing what was stored on line 487. Is that intentional?

Reply:
See above. All command use the same template. It will be factorized.

Thanks for all your comments